File: dotnet/src/PqcStandards/SlhDsa/Address.cs
namespace PqcStandards.SlhDsa;

/// <summary>ADRS (address) structure for SLH-DSA. 32 bytes, 7 types.</summary>
public class Address
{
    public const int WotsHash = 0;
    public const int WotsPk = 1;
    public const int TreeNode = 2;
    public const int ForsTree = 3;
    public const int ForsPk = 4;
    public const int WotsCompress = 5;
    public const int ForsCompress = 6;

    private readonly byte[] _data = new byte[32];

    public byte[] Data => _data;

    public void SetLayerAddress(int layer)
    {
        _data[0] = (byte)(layer >> 24);
        _data[1] = (byte)(layer >> 16);
        _data[2] = (byte)(layer >> 8);
        _data[3] = (byte)layer;
    }

    public void SetTreeAddress(long tree)
    {
        // Bytes 4-7: zero; bytes 8-15: uint64 big-endian (FIPS 205 Figure 2)
        _data[4] = 0; _data[5] = 0; _data[6] = 0; _data[7] = 0;
        _data[8]  = (byte)(tree >> 56);
        _data[9]  = (byte)(tree >> 48);
        _data[10] = (byte)(tree >> 40);
        _data[11] = (byte)(tree >> 32);
        _data[12] = (byte)(tree >> 24);
        _data[13] = (byte)(tree >> 16);
        _data[14] = (byte)(tree >> 8);
        _data[15] = (byte)tree;
    }

    public void SetType(int type)
    {
        // Bytes 16-19: type; bytes 20-31: zeroed (FIPS 205 Figure 2)
        _data[16] = (byte)(type >> 24);
        _data[17] = (byte)(type >> 16);
        _data[18] = (byte)(type >> 8);
        _data[19] = (byte)type;
        Array.Clear(_data, 20, 12);
    }

    public void SetKeyPairAddress(int kp)
    {
        _data[20] = (byte)(kp >> 24);
        _data[21] = (byte)(kp >> 16);
        _data[22] = (byte)(kp >> 8);
        _data[23] = (byte)kp;
    }

    public int GetKeyPairAddress()
    {
        return (_data[20] << 24) | (_data[21] << 16) | (_data[22] << 8) | _data[23];
    }

    public void SetChainAddress(int chain)
    {
        _data[24] = (byte)(chain >> 24);
        _data[25] = (byte)(chain >> 16);
        _data[26] = (byte)(chain >> 8);
        _data[27] = (byte)chain;
    }

    public void SetHashAddress(int hash)
    {
        _data[28] = (byte)(hash >> 24);
        _data[29] = (byte)(hash >> 16);
        _data[30] = (byte)(hash >> 8);
        _data[31] = (byte)hash;
    }

    public void SetTreeHeight(int height)
    {
        _data[24] = (byte)(height >> 24);
        _data[25] = (byte)(height >> 16);
        _data[26] = (byte)(height >> 8);
        _data[27] = (byte)height;
    }

    public void SetTreeIndex(int index)
    {
        _data[28] = (byte)(index >> 24);
        _data[29] = (byte)(index >> 16);
        _data[30] = (byte)(index >> 8);
        _data[31] = (byte)index;
    }

    public Address Copy()
    {
        var a = new Address();
        Buffer.BlockCopy(_data, 0, a._data, 0, 32);
        return a;
    }
}

File: php/src/SlhDsa/Address.php
<?php

declare(strict_types=1);

namespace PQC\SlhDsa;

/**
 * ADRS (Address) structure for SLH-DSA.
 * 32-byte address used to domain-separate hash calls.
 */
final class Address
{
    // Address types
    public const WOTS_HASH = 0;
    public const WOTS_PK = 1;
    public const TREE = 2;
    public const FORS_TREE = 3;
    public const FORS_ROOTS = 4;
    public const WOTS_PRF = 5;
    public const FORS_PRF = 6;

    private string $data;

    public function __construct()
    {
        $this->data = str_repeat("\x00", 32);
    }

    public function copy(): self
    {
        $new = new self();
        $new->data = $this->data;
        return $new;
    }

    public function toBytes(): string
    {
        return $this->data;
    }

    /**
     * Set layer address (bytes 0-3).
     */
    public function setLayerAddress(int $layer): self
    {
        $this->setWord(0, $layer);
        return $this;
    }

    /**
     * Set tree address (bytes 4-15).
     */
    public function setTreeAddress(int $tree): self
    {
        // FIPS 205 Figure 2: bytes 4-7 zero, bytes 8-15 hold uint64 big-endian
        $this->data[4]  = "\x00";
        $this->data[5]  = "\x00";
        $this->data[6]  = "\x00";
        $this->data[7]  = "\x00";
        $this->data[8]  = chr(($tree >> 56) & 0xFF);
        $this->data[9]  = chr(($tree >> 48) & 0xFF);
        $this->data[10] = chr(($tree >> 40) & 0xFF);
        $this->data[11] = chr(($tree >> 32) & 0xFF);
        $this->data[12] = chr(($tree >> 24) & 0xFF);
        $this->data[13] = chr(($tree >> 16) & 0xFF);
        $this->data[14] = chr(($tree >> 8) & 0xFF);
        $this->data[15] = chr($tree & 0xFF);
        return $this;
    }

    /**
     * Set address type (bytes 16-19).
     */
    public function setType(int $type): self
    {
        $this->setWord(16, $type);
        // Reset subsequent fields when type changes
        $this->data[20] = "\x00";
        $this->data[21] = "\x00";
        $this->data[22] = "\x00";
        $this->data[23] = "\x00";
        $this->data[24] = "\x00";
        $this->data[25] = "\x00";
        $this->data[26] = "\x00";
        $this->data[27] = "\x00";
        $this->data[28] = "\x00";
        $this->data[29] = "\x00";
        $this->data[30] = "\x00";
        $this->data[31] = "\x00";
        return $this;
    }

    /**
     * Set keypair address (bytes 20-23).
     */
    public function setKeyPairAddress(int $kp): self
    {
        $this->setWord(20, $kp);
        return $this;
    }

    /**
     * Set chain address (bytes 24-27).
     */
    public function setChainAddress(int $chain): self
    {
        $this->setWord(24, $chain);
        return $this;
    }

    /**
     * Set hash address (bytes 28-31).
     */
    public function setHashAddress(int $hash): self
    {
        $this->setWord(28, $hash);
        return $this;
    }

    /**
     * Set tree height (bytes 24-27).
     */
    public function setTreeHeight(int $height): self
    {
        $this->setWord(24, $height);
        return $this;
    }

    /**
     * Set tree index (bytes 28-31).
     */
    public function setTreeIndex(int $index): self
    {
        $this->setWord(28, $index);
        return $this;
    }

    /**
     * Get keypair address.
     */
    public function getKeyPairAddress(): int
    {
        return $this->getWord(20);
    }

    /**
     * Set 4 bytes at offset as big-endian 32-bit integer.
     */
    private function setWord(int $offset, int $value): void
    {
        $this->data[$offset] = chr(($value >> 24) & 0xFF);
        $this->data[$offset + 1] = chr(($value >> 16) & 0xFF);
        $this->data[$offset + 2] = chr(($value >> 8) & 0xFF);
        $this->data[$offset + 3] = chr($value & 0xFF);
    }

    /**
     * Get 4-byte big-endian value.
     */
    private function getWord(int $offset): int
    {
        return (ord($this->data[$offset]) << 24) |
               (ord($this->data[$offset + 1]) << 16) |
               (ord($this->data[$offset + 2]) << 8) |
               ord($this->data[$offset + 3]);
    }
}
